doc/source/development.rst

```diff
@@ -23,14 +23,14 @@ you can try adding ``--user``. You may also need to run something like ``sudo
 chown -R $USER /home/ubuntu/anaconda3`` (substituting in the appropriate
 path).
 
-If you make changes to the C++ files, you will need to recompile them.
+If you make changes to the C++ or Python files, you will need to run the build so C++ code is recompiled and/or Python files are redeployed in `ray/python`.
 However, you do not need to rerun ``pip install -e .``. Instead, you can
 recompile much more quickly by doing
 
 .. code-block:: shell
 
  cd ray
- bazel build //:ray_pkg
+ bash build.sh
```

Comment:
Can we replace this with something like
```
PYTHON_BIN_PATH=`which python` bazel build //:ray_pkg
```?

We are planning to get rid of `build.sh` really soon :)

Reply:
It would be great to keep build.sh so the process of building is encapsulated in script instead of everyone needing to shuffle through docs and copy paste commands. If these command changes then you just need to update build.sh instead of broadcasting fix manually to everyone. I also strongly believe there should be install.sh instead of people having to copy past half dozen commands every single time.

Reply:
If you want to compile Ray from source, the official way to do it is
```
pip install -e .
```
inside the `ray/python/` folder. Does that not work for you? If it doesn't, we should fix it.

If you are a developer and want super high turn around time for recompiling, you should run the above command once and then
```
PYTHON_BIN_PATH=`which python` bazel build //:ray_pkg
```
for very fast incremental builds. But really `pip install -e .` is often fast enough in this case too.

`build.sh` is still remaining from the old days when Ray used `CMake` as a build system and a lot of more setup was needed (like setting up third party libraries). But nowadays with bazel, it is not needed any more.

Reply:
pip works and install is not a problem at all. However once I change source then rebuilding is an issue. I thought build.sh does rebuilding as it is essentially running above command. In most projects I work with, they have install.sh for installing and build.sh for rebuilding code if you have made edits. This avoids additional burden on people having to learn all these details.

I really think we should avoid squeeze complexity into one liners like above. Things might change in future and people shouldn’t have to relearn every time how to rebuild. Also, one liners are like above are not complete. For example, you might need PYTHON_LIB_PATH as well for bazel at some point. It’s just nice to encapsulate these stuff in to build.sh or rebuild.sh. 

Similarly, pip is not enough to install rat. You also need to build bazel, install packages like tabulate and filelock and do few more things for dashboard. I had to manually copy paste half dozen statements to get through this. Imagine every rat user having to do this every single time. It would great to encapsulate al, these statements simply in install.sh.

Reply:
We can make `pip install -e .` really fast and not do additional work for a rebuild. One reason we are trying to get rid of `build.sh` is because we are porting Ray to windows and shell scripts don't work well for that.

For installing Ray, `pip` is also the way to go. It will install all dependenies for you too.

Reply:
For complex projects, pip often won’t be sufficient. For ray, for example, dashboard requires whole bunch of other steps. I think of build.sh and install.sh as verbs for the project so I don’t need to worry about all these and just issue the verbs. For me, build.sh is for rebuilding after source changes. For Windows, you can have build.cmd and install.cmd. This does increase duplication but it simplifies things for users.

Reply:
Fair enough, I think you convinced me. We can plan to have a `dev/build.py` that also can build the dashboard.